[FLINK-1319][core] Add static code analysis for UDFs

twalthr opened the pull request:
This PR implements a Static Code Analyzer (SCA) that uses the ASM framework for interpreting Java bytecode of Flink UDFs. The analyzer is build on top of ASM's `BasicInterpreter`. Instead of ASM's `BasicValue`s, I introduced `TaggedValue`s which extends `BasicValue` and allows for appending interesting information to values. Interesting values such as inputs, collectors, or constants are tagged such that a tracking of atomic input fields through the entire UDF (until the function returns or calls `collect()`) is possible.

The implementation is as conservative as possible meaning that for cases or bytecode instructions that haven't been considered the analyzer will fallback to the ASM library (which removes TaggedValues).

61 JUnit tests are testing the basic functionality. 18 JUnit tests with code examples from the "real world" are testing the analyzer even more.

The analyzer has 3 modes: DISABLED, OPTIMIZE, HINTS

The interpretation takes some time. It is possible that an analysis of an UDF takes up to 1 second. Therefore, I didn't enable the analyzer in TestEnvironment by default to reduce the build times, but if you uncomment the lines the analyzer supports all 280 UDFs within the entire Flink code. 

The analyzer gives hints about:
- Main feature: ForwardedFields semantic properties for all types of Functions except for MapPartition and Combine
- Warnings if static fields are modified by a Function
- Warnings if a FilterFunction modifies its input objects
- Warnings if a Function returns `null`
- Warnings if a tuple access uses a wrong index
- Information about the number of object creations within a UDF (for manual optimization)

rmetzger commented:
Thanks a lot for this great work!
I'll soon try out the code to see how it works.

Given that we can disable it and that its not automatically setting semantic properties, I would vote to merge it soon to include it into the 0.9 release.

StephanEwen commented:
Indeed, this looks like an impressive addition.

Let's get it into 0.9 as a beta feature!

StephanEwen commented:
Looks very nice, seems to have a good test coverage as well.

How well does it work with bytecode generated by the Scala Compiler?

StephanEwen commented:
The tests are failing with checkstyle violations. Can you fix those?

twalthr commented:
The bytecode generated from Scala Compiler is the same. That is not a problem for the analyzer. But Scala is not fully supported yet, because of the different Java/Scala Tuples (fields starting with "_" instead of "f" etc.). I will add support for that in the near future.

If any exceptions are thrown, they are only visible in the debug log. So we could merge it into the 0.9 release without any disadvantages.

It builds now ;)

StephanEwen commented:
Okay, pending a proper distributed test, you have my vote to add this!

uce commented:
This is a very cool! Thanks for all the effort you put into this. I think this will be a great addition to the project. :-)

---

I've skimmed over the code and checked it out. Some initial feedback:
- **Naming UDF**. The feedback of committers giving talks about Flink some time ago was that the name "UDF" was sometimes confusing. @rmetzger, can you confirm this? We might take this into account and rename the `UdfAnalysisMode` to something else, for example just `CodeAnalysisMode`.
- **Default mode**: the default mode is currently set to hinting. Running the WordCount example, it didn't work out of the box though. After some debugging I found this in the `UdfAnalyzer`, which prevented the analyis:
  ```java
if (internalUdfClassName.startsWith(EXCLUDED_CLASSPATH)
        && !internalUdfClassName.startsWith("org/apache/flink/api/java/sca")) {
        return false;
}
```
  After commenting the second condition out, it worked fine. For a WordCount with a reducer, which only touches the second field of the Tuple2, it worked like a charm. The produced output looks like this:

  ```
23:23:49,207 INFO  org.apache.flink.api.java.operators.UdfOperatorUtils          - Function 'org.apache.flink.examples.java.wordcount.WordCount$Tokenizer' has been analyzed with the following result:
Number of object creations (should be kept to a minimum): 1 in method / 36 transitively
A need for forwarded fields annotations could not be found.

23:23:49,217 INFO  org.apache.flink.api.java.operators.UdfOperatorUtils          - Function 'org.apache.flink.examples.java.wordcount.WordCount$1' has been analyzed with the following result:
Number of object creations (should be kept to a minimum): 1 in method / 1 transitively
You could use the following annotation: @ForwardedFields("f0->f0;")

23:23:49,243 INFO  org.apache.flink.api.java.operators.UdfOperatorUtils          - Function 'org.apache.flink.api.java.Utils$CollectHelper' has been analyzed with the following result:
Number of object creations (should be kept to a minimum): 0 in method / 6 transitively
A need for forwarded fields annotations could not be found.
```
  Very nice to see this working. This is great news for the optimizer. :-)

  What do the transitive object creations refer to exactly? I'm wondering how a user could influence them? Reusing a result object in the WordCount sum reducer is correctly detected as 0 creations in the method as well.

  I was wondering whether it could be possible to configure code analysis on a per-function level. For example, all library related functions should not print the hints imo.

---

I very much like this. Regarding merging this before the release: I would vote to only merge it if we disable it by default. This will essentially affect every program written in the upcoming Flink release and merging it without proper review, testing, and exposure seems rushed to me. But I wouldn't veto it, if everyone wants it in asap.

I think this feature alone would warrant a new 0.10 release soon after 0.9.

StephanEwen commented:
I second Ufuk's comments.
Merging it and deactivating it by default. I can see a 0.9.1 or 0.10.0 release coming in very soon afterwards, because we have a big set of issues still in the pipeline.

Initially activating hinting in the local environment (what people use during debigging anyways) and having it deactivated in the "production" environments (remote and context).

Other comments:
  - How about printing the hints to sysout? I can see them getting lost among the logging statements. Also, people often have logging not activated in the IDE.
  - Package based exclusions never worked, it was always an issue with the quickstarts. I assume you want the exclusion to make sure you do not analyze the built-in default join function, for example? What you can do is add an annotation that says "DoNotAnalyze" to that functions, and then simply analyze everything.

twalthr commented:
I disabled the analyzer for all classes starting with "org.apache.flink". Because I wanted to reduce the output to the user for build-in UDFs (e.g. `org.apache.flink.api.java.Utils$CollectHelper` or UDFs within the Graph API). Initially I thought about an annotation `@SkipCodeAnalysis` but there are too many UDFs where this annotation should then be placed at. I think we can assume that UDFs shipped with Flink are already implemented effcient or unefficient for example purposes only.

Object creations "in method" mean that these objects are created directly in e.g. `map()`. The analyzer also follows method calls. "transitively" created objects are objects created in the nested method calls.

twalthr commented:
+1 for activating hinting locally. 
I thought logging is the standard way to print, but I can change it to a sysout

uce commented:
- I agree with hints going to sysout and activating by default.
- For simple functions, most of the transitive allocations will be Flink internal, right? E.g. after calling the collect method. Would it make sense to exclude transitive allocations reached via collect?

twalthr commented:
The `collect()` is a special case, the analyzer does not follow it. But after thinking about it I recognized that I forgot `getRuntimeContext()` I will fix this ;)

StephanEwen commented:
Let us merge this for 0.9 and have it deactivated by default.

Let's gradually activate it in the next releases as it gets exposure

uce commented:
OK, I will review this.

I vote to stick to Stephan's suggested approach instead of package based exclusions: analyze everything and allow exclusions with a `@SkipCodeAnalysis` annotation.

Any further opinions on the output of the analysis (stdout vs. logging question)?

rmetzger commented:
+1 for the annotation.

I'm against using stdout. Logging frameworks are much better at controlling the output.
The quickstart mvn archetype provides a log4j.properties file, so we can configure it the way we want it to.

StephanEwen commented:
The concern about logging is that, when using the local mode inside the IDE, the system logs a lot and the hints get lost.

If you don't want sysoutput, you could always deactivate the analysis.

rmetzger commented:
I understand that concern.

But using sysoutput will also be interleaved with the Client sysout printing and the regular system logging.
Also, I think its very bad practice to print stuff using systemout, because its not controllable in any way.
With log4j we can configure the analysis output the way we want.
if you want the messages to look like regular sysout text, we can specify a custom output schema for the classes in the sca java package.

rmetzger commented:
> Naming UDF. The feedback of committers giving talks about Flink some time ago was that the name "UDF" was sometimes confusing. @rmetzger, can you confirm this? We might take this into account and rename the UdfAnalysisMode to something else, for example just CodeAnalysisMode.

That's right. People associate UDFs with SQL databases that allow to pass in custom functions (which is right, but they start thinking Flink is a SQL database).
In this case, its not super critical because its internal code.

uce commented on flink-core/src/main/java/org/apache/flink/api/common/UdfAnalysisMode.java:
```diff
@@ -0,0 +1,45 @@
+ * the user about common implementation mistakes.
+ *
+ */
+public enum UdfAnalysisMode {
```
This is user-facing. I vote to rename it. @rmetzger agrees that in his experience the UDF part can be misleading. I understand why you chose this though... the operators make use of the UDF term all over the place. What about `CodeAnalysisMode`? After all both this PR and the package are called *code analysis* and not UDF analysis.

uce commented on flink-core/src/main/java/org/apache/flink/api/common/UdfAnalysisMode.java:
```diff
@@ -0,0 +1,45 @@
+	 */
+	OPTIMIZING_ENABLED;
+
+}
```
Since the user will have to set this, what about keeping it short? `DISABLE`, `HINT`, `OPTIMIZE`?

uce commented on flink-core/src/main/java/org/apache/flink/api/common/UdfAnalysisMode.java:
```diff
@@ -0,0 +1,45 @@
+package org.apache.flink.api.common;
+
+/**
+ * Specifies to which extent user-defined functions are analyzed in order
```
- I would make this more concrete. What about the list from your initial PR comment?
- line 25: empty

uce commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SingleInputUdfOperator.java:
```diff
@@ -54,8 +54,11 @@
 
 	private Map<String, DataSet<?>> broadcastVariables;
 
+	// NOTE: only set this variable via setSemanticProperties()
```
I think this refactoring is quite fragile. The semantic properties utility is not returning an empty properties object, but null and you take care of setting it correctly here depending on whether the forwarded fields have been set manually or not.

If optimize is enabled and there are manual annotations, they will be overriden. I am wondering if it is better to have manual annotations trump optimizer annotations. What's your opinion on this?

uce replied:
I saw that there is a check for this in the UdfAnalyzerUtil.

twalthr replied:
Manual annotations should always trump optimizer annotations. The analyzer can not determine all semantic properties. E.g. when using KeySelectors. The user should still have the possibility to override semantic properties to add more properties.

fhueske replied:
Just a quick question (haven't checked the code). Does the analyzer also respect semantic information provided via the Operator API (withForwardedFields()), i.e., not via function annotations?

twalthr replied:
Yes, I have also added a test for that (see SemanticPropertiesPrecendenceTest).

uce commented on flink-java/src/main/java/org/apache/flink/api/java/operators/UdfOperatorUtils.java:
```diff
@@ -0,0 +1,94 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(UdfOperatorUtils.class);
+
+	public static void analyzeSingleInputUdf(SingleInputUdfOperator<?, ?, ?> operator, Class<?> udfBaseClass,
```
I vote to pass the name of the operator as well. The log output will then be more consistent. Currently the class name is used.

uce commented on flink-java/src/main/java/org/apache/flink/api/java/sca/UdfAnalyzer.java:
```diff
@@ -0,0 +1,431 @@
+
+public class UdfAnalyzer {
+	// exclusion to suppress hints for API operators
+	private static final String EXCLUDED_CLASSPATH = "org/apache/flink";
```
Instead of excluding this class path, the consensus from reviews so far is to add an `@SkipCodeAnalysis` annotation. This will allow new users to play around with the Flink examples etc.

uce commented on flink-java/src/main/java/org/apache/flink/api/java/operators/UdfOperatorUtils.java:
```diff
@@ -0,0 +1,94 @@
+					if (mode == UdfAnalysisMode.OPTIMIZING_ENABLED
+							&& !operator.udfWithForwardedFieldsAnnotation(udf.getClass())) {
+						operator.setSemanticProperties((SingleInputSemanticProperties) analyzer.getSemanticProperties());
+						operator.setAnalyzedUdfSemanticsFlag();
```
I think it would make sense to also print the inferred forwarded fields (at least for debugging purposes).

uce commented on flink-java/src/main/java/org/apache/flink/api/java/operators/UdfOperatorUtils.java:
```diff
@@ -0,0 +1,94 @@
+	public static void analyzeDualInputUdf(TwoInputUdfOperator<?, ?, ?, ?> operator, Class<?> udfBaseClass,
+			Function udf, Keys<?> key1, Keys<?> key2) {
+		final UdfAnalysisMode mode = operator.getExecutionEnvironment().getConfig().getUdfAnalysisMode();
+		if (mode != UdfAnalysisMode.DISABLED) {
```
We could log that the analysis is disabled as well.

uce commented:
Hey Timo,

I've tested this locally so far and it's working smoothly! Let's write a blog post about this very soon. :-) As soon as I have access to more machines, I will test it on a cluster. The stuff denoted with [RB] should be fixed in any case before merging imo. The rest we can also do afterwards.

**USER FACING**

1. Disable the analysis by default. [RB]

2. This is cumbersome, but we should go for Annotation-based exclusions instead of package based (see inline comments). [RB]

3. Currently manual annotations trump automatic ones, which is a good thing, because people won't have unexpected results. Could you add a test for this?

4. I would give the analyzer util the call location based function name (String callLocation = Utils.getCallLocationName()). I think that will have better output than just `Function 'Job$2' has been analyzed with the following result: ...`

4. I would rename `UdfAnalyisMode` (see inline comments).

5. The hints are currently logged on a singe line w/o a whitespace after each hint, like this:
   ```
Function modifies static fields. This can lead to unexpected behaviour during runtime.Function returns 'null' values. This can lead to errors during runtime.A need for forwarded fields annotations could not be found.
```
6. Some analysis like a wrong tuple index access or returning null lead fail the program before submitting it, which is very nice. I would actually like to log these problems on a log level WARN (instead of INFO) when only hints are enabled.

7. When you have optimize enabled and run a filter, which changes the input, the error you get is this. You can't tell what's wrong.
    ```
Exception in thread "main" org.apache.flink.api.java.sca.UdfErrorException: UDF contains obvious errors.
	at org.apache.flink.api.java.sca.UdfAnalyzer.analyze(UdfAnalyzer.java:300)
	...
```
    For a wrong tuple index access, it is correct:
    ```
Exception in thread "main" org.apache.flink.api.java.sca.UdfErrorException: UDF contains obvious errors.
	at org.apache.flink.api.java.sca.UdfAnalyzer.analyze(UdfAnalyzer.java:300)
	...
Caused by: org.apache.flink.api.java.sca.UdfErrorException: Function contains tuple accesses with invalid indexes. This can lead to errors during runtime.
	at org.apache.flink.api.java.sca.UdfAnalyzer.addHintOrThrowException(UdfAnalyzer.java:413)
	...
```
    I'm not sure if the thrown Execption should be an `UdfErrorException` or an `UdfAnalysisException`?
8. When optimizing is enabled, I would still print the inferred forwarded fields.

9. Regarding the result messages: personally, I think the hints/messages could be more consise, e.g. for example skip the `(should be kept to a minimum)` in the number of object creations msg or just say `Forwarded fields: none` instead of `A need for forwarded fields annotations could not be found.`

**INTERNALS**

1. I think the internals could use more comments. It's easy to get the general idea, but it would be nice to also get some high-level comments about the ASM-related stuff. I didn't have time to dig deep into it.

---

Will report back after cluster tests.

rmetzger commented:
How about extending the `UDF contains obvious errors` message with some notes on how to completely disable the SCA.
I fear that the message appears (blocks the program execution) due to a bug in the SCA and then users don't know how to get their stuff to run.

uce commented:
I agree with Robert, but for the initial version it won't matter as it should be disabled anyways.

StephanEwen commented:
Great review, Ufuk.

I agree with @uce and @rmetzger to add a comment how to disable it (in case).

twalthr commented:
Hey Ufuk,

thank you very much for reviewing my code and all others for the feedback! I tried to consider all your feedback (I hope I didn't forget anything). I did a large refactoring again, added some comments to important parts of the code and fixed some bugs. I also added some additional test cases. I hope the PR is now ready to be merged (if the build succeeds) :)

uce commented:
The build succeeds. :) I will have a look at the changes. Thanks for not force updating this PR.

I will test it in a distributed setup and if everything runs fine, we can merge this. :-)

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/functions/SemanticPropUtil.java:
```diff
@@ -309,15 +308,20 @@ public static DualInputSemanticProperties getSemanticPropsDual(
-			TypeInformation<?> inType, TypeInformation<?> outType)
-	{
+			TypeInformation<?> inType, TypeInformation<?> outType,
+			boolean skipIncompatibleTypes) {
```
Can you explain a bit, why you introduced this flag? 
Why should is be possible to skip the compatibility checks?

twalthr replied:
Sometimes the analyzer works better than required. E.g. the analyzer outputs @ForwardedFields("*->record.customer.name") but if customer is a GenericType output type, the types are incompatible. I thought it is better to reuse the type compatibility checking of the PropUtil than reimplement everything, but skip types that are incompatible without throwing Exceptions.

fhueske replied:
OK, got it. Thanks for explaining. :-)

uce commented:
Thanks for the changes. :-) They look very good. Cluster tests are also fine.

---

I would like to address the following points, squash your commits and merge it. Is this fine with you?

1. Add the function name to the CodeErrorExceptions. Currently you just get

    ```
Exception in thread "main" org.apache.flink.api.java.sca.CodeErrorException: Function code contains obvious errors. If you think the code analysis is wrong at this point you can disable the entire code analyzer in ExecutionConfig or add @SkipCodeAnalysis to your function to disable the analysis.
	at org.apache.flink.api.java.sca.UdfAnalyzer.analyze(UdfAnalyzer.java:328)
	at org.apache.flink.api.java.operators.UdfOperatorUtils.analyzeSingleInputUdf(UdfOperatorUtils.java:46)
	at org.apache.flink.api.java.operators.FilterOperator.<init>(FilterOperator.java:45)
	at org.apache.flink.api.java.DataSet.filter(DataSet.java:282)
	at Job.main(Job.java:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: org.apache.flink.api.java.sca.CodeErrorException: Function contains tuple accesses with invalid indexes. This can lead to errors during runtime.
```
    but it is important to see, which function exactly failed.
2. We now print the hint string in `OPTIMIZE` mode, but I think we should make it even more explicit when an optimization is applied.

twalthr commented:
Good points. Perfectly fine for me ;)

uce commented:
Travis is running... https://github.com/uce/incubator-flink/tree/sca :)

uce commented:
Initial version is merged in DISABLE mode! Thanks for the contribution. I will file some follow up JIRAs. :)

twalthr commented:
Great news :) Thanks Ufuk!
